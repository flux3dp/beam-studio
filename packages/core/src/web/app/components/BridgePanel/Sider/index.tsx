import type { Dispatch, SetStateAction } from 'react';
import React, { memo } from 'react';

import { QuestionCircleOutlined } from '@ant-design/icons';
import { Button, Flex, Form, Switch } from 'antd';

import BridgePanelIcons from '@core/app/icons/BridgePanel/BridgePanelIcons';
import BackButton from '@core/app/widgets/FullWindowPanel/BackButton';
import Footer from '@core/app/widgets/FullWindowPanel/Footer';
import Header from '@core/app/widgets/FullWindowPanel/Header';
import FullWindowPanelSider from '@core/app/widgets/FullWindowPanel/Sider';
import UnitInput from '@core/app/widgets/UnitInput';
import useI18n from '@core/helpers/useI18n';

import styles from './index.module.scss';

interface Props {
  bridgeGap: number;
  bridgeWidth: number;
  handleCutPathByGap: () => void;
  mode: 'auto' | 'manual';
  onClose: () => void;
  onComplete: () => void;
  setBridgeGap: Dispatch<SetStateAction<number>>;
  setBridgeWidth: Dispatch<SetStateAction<number>>;
  setMode: Dispatch<SetStateAction<'auto' | 'manual'>>;
}

function UnmemorizedSider({
  bridgeGap,
  bridgeWidth,
  handleCutPathByGap,
  mode = 'manual',
  onClose,
  onComplete,
  setBridgeGap,
  setBridgeWidth,
  setMode,
}: Props): React.JSX.Element {
  const { buttons: langButtons } = useI18n();

  return (
    <FullWindowPanelSider className={styles.sider}>
      <Flex className={styles['h-100']} justify="space-between" vertical>
        <div>
          <BackButton onClose={onClose}>{langButtons.back_to_beam_studio}</BackButton>
          <Header icon={<BridgePanelIcons.Bridge />} title={'Bridge'} />
          <div className={styles.wrapper}>
            <Form>
              <Form.Item label={`Manual Mode:`}>
                <>
                  <Switch
                    className={styles.switch}
                    onChange={(checked) => setMode(checked ? 'manual' : 'auto')}
                    value={mode === 'manual'}
                  />
                  <QuestionCircleOutlined className={styles.icon} />
                </>
              </Form.Item>
              <Form.Item hidden={mode !== 'manual'} label={`Width:`}>
                <UnitInput
                  addonAfter="mm"
                  data-testid="bridge-width"
                  isInch={false}
                  key="bridge-width"
                  max={10}
                  min={0.1}
                  onChange={(value) => setBridgeWidth(value!)}
                  precision={1}
                  step={1}
                  type="number"
                  value={bridgeWidth}
                />
              </Form.Item>
              <Form.Item hidden={mode !== 'auto'} label={`Gap:`}>
                <UnitInput
                  addonAfter="mm"
                  data-testid="bridge-gap"
                  isInch={false}
                  key="bridge-gap"
                  max={50}
                  min={1}
                  onChange={(value) => setBridgeGap(value!)}
                  precision={1}
                  step={1}
                  value={bridgeGap}
                />
              </Form.Item>
              <Button hidden={mode !== 'auto'} onClick={handleCutPathByGap} type="default">
                Add Bridge
              </Button>
            </Form>
          </div>
        </div>
        <Footer>
          <Button key="ok" onClick={onComplete} type="primary">
            {'Save and Exit'}
          </Button>
        </Footer>
      </Flex>
    </FullWindowPanelSider>
  );
}

const Sider = memo(UnmemorizedSider);

export default Sider;
