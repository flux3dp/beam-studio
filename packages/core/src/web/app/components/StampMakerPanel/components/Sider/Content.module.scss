.wrapper {
  height: 100%;
  flex: 1;
  padding: 20px 40px 0;
  overflow: hidden;

  @media screen and (max-width: 600px) {
    padding: 12px 12px 0;
  }

  .tool-tip {
    color: #7c7c7c;
    cursor: help;
  }
}
