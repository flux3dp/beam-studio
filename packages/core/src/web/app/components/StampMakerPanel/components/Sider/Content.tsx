import React, { useMemo } from 'react';

import { QuestionCircleOutlined } from '@ant-design/icons';
import { Flex, Form, Switch } from 'antd';
import Konva from 'konva';
import { match } from 'ts-pattern';

import UnitInput from '@core/app/widgets/UnitInput';
import useI18n from '@core/helpers/useI18n';

import { useStampMakerPanelStore } from '../../store';

import styles from './Content.module.scss';

export default function Content(): React.JSX.Element {
  const { stamp_maker_panel } = useI18n();
  const { addFilter, bevelRadius, filters, horizontalFlip, removeFilter, setBevelRadius, setHorizontalFlip } =
    useStampMakerPanelStore();
  const isInverted = useMemo(() => filters.includes(Konva.Filters.Invert), [filters]);

  const handleToggleFlip = () => setHorizontalFlip(!horizontalFlip);
  const handleToggleInvert = () =>
    match(isInverted)
      .with(true, () => removeFilter(Konva.Filters.Invert))
      .otherwise(() => addFilter(Konva.Filters.Invert, true));

  return (
    <div className={styles.wrapper}>
      <Form layout="horizontal">
        <Form.Item
          label={stamp_maker_panel.invert}
          tooltip={{ icon: <QuestionCircleOutlined />, title: stamp_maker_panel.tool_tip.invert }}
        >
          <Switch checked={isInverted} onChange={handleToggleInvert} />
        </Form.Item>
        <Form.Item
          label={stamp_maker_panel.horizontal_flip}
          tooltip={{ icon: <QuestionCircleOutlined />, title: stamp_maker_panel.tool_tip.horizontal_flip }}
        >
          <Switch checked={horizontalFlip} onChange={handleToggleFlip} />
        </Form.Item>
        <Form.Item label={stamp_maker_panel.bevel_radius} layout="vertical">
          <Flex justify="space-between">
            <UnitInput
              //
              addonAfter="mm"
              max={10}
              min={0}
              onChange={(value) => {
                if (value !== null) {
                  setBevelRadius(value);
                }
              }}
              step={0.1}
              value={bevelRadius}
            />
          </Flex>
        </Form.Item>
      </Form>
    </div>
  );
}
