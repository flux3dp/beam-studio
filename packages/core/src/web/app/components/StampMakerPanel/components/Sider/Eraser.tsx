import React from 'react';

import { QuestionCircleOutlined } from '@ant-design/icons';
import { Flex, Form, InputNumber, Slider } from 'antd';

import useI18n from '@core/helpers/useI18n';

import styles from './PanelContent.module.scss';

const MAX_BRUSH_SIZE = 128;

export default function Content(): React.JSX.Element {
  const { image_edit_panel: lang } = useI18n();

  return (
    <div className={styles.wrapper}>
      <div className={styles.title}>{lang.eraser.title}</div>
      <div className={styles['hint-text']}>
        <QuestionCircleOutlined className={styles.icon} />
        <span>{lang.eraser.description}</span>
      </div>
      <Form layout="vertical">
        <Form.Item label={`${lang.eraser.brush_size}:`}>
          <Flex align="center" gap={8} justify="between">
            <Slider className={styles.slider} max={MAX_BRUSH_SIZE} min={1} onChange={() => {}} step={1} value={20} />
            <InputNumber
              max={MAX_BRUSH_SIZE}
              min={1}
              onChange={(val) => {
                if (val !== null) return;
              }}
              step={1}
              value={20}
            />
          </Flex>
        </Form.Item>
      </Form>
    </div>
  );
}
