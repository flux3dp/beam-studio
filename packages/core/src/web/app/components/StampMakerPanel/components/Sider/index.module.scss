.sider {
  width: 356px;
  min-width: 356px;
}

.icon {
  font-size: 24px;
}

.h-100 {
  height: 100%;
}
