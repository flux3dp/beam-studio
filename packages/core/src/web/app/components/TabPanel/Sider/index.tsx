import type { Dispatch, SetStateAction } from 'react';
import React, { memo } from 'react';

import { Button, Flex, Form, Switch } from 'antd';

import TabPanelIcons from '@core/app/icons/TabPanel/TabPanelIcons';
import BackButton from '@core/app/widgets/FullWindowPanel/BackButton';
import Footer from '@core/app/widgets/FullWindowPanel/Footer';
import Header from '@core/app/widgets/FullWindowPanel/Header';
import FullWindowPanelSider from '@core/app/widgets/FullWindowPanel/Sider';
import UnitInput from '@core/app/widgets/UnitInput';
import useI18n from '@core/helpers/useI18n';
import browser from '@core/implementations/browser';

import styles from './index.module.scss';

interface Props {
  gap: number;
  handleCutPathByGap: () => void;
  mode: 'auto' | 'manual';
  onClose: () => void;
  onComplete: () => void;
  setGap: Dispatch<SetStateAction<number>>;
  setMode: Dispatch<SetStateAction<'auto' | 'manual'>>;
  setWidth: Dispatch<SetStateAction<number>>;
  width: number;
}

function UnmemorizedSider({
  gap,
  handleCutPathByGap,
  mode = 'manual',
  onClose,
  onComplete,
  setGap,
  setMode,
  setWidth,
  width,
}: Props): React.JSX.Element {
  const { alert, buttons: langButtons, global, tab_panel: tabPanel } = useI18n();

  return (
    <FullWindowPanelSider className={styles.sider}>
      <Flex className={styles['h-100']} justify="space-between" vertical>
        <div>
          <BackButton onClose={onClose}>{langButtons.back_to_beam_studio}</BackButton>
          <Header icon={<TabPanelIcons.Tab />} title={tabPanel.title} />
          <div className={styles.wrapper}>
            <div className={styles.link} onClick={() => browser.open(tabPanel.help_center)}>
              {alert.learn_more}
            </div>
            <Form>
              <Form.Item label={`${tabPanel.manual_mode}:`}>
                <Switch
                  className={styles.switch}
                  onChange={(checked) => setMode(checked ? 'manual' : 'auto')}
                  value={mode === 'manual'}
                />
              </Form.Item>
              <Form.Item label={`${tabPanel.width}:`}>
                <UnitInput
                  addonAfter="mm"
                  changeOnWheel
                  fireOnChange
                  isInch={false}
                  key="width"
                  max={10}
                  min={0.1}
                  onChange={(value) => setWidth(value!)}
                  precision={1}
                  step={0.1}
                  type="number"
                  value={width}
                />
              </Form.Item>
              <Form.Item hidden={mode !== 'auto'} label={`${tabPanel.gap}:`}>
                <UnitInput
                  addonAfter="mm"
                  changeOnWheel
                  fireOnChange
                  isInch={false}
                  key="gap"
                  max={100}
                  min={1}
                  onChange={(value) => setGap(value!)}
                  precision={1}
                  step={1}
                  value={gap}
                />
              </Form.Item>
              <Button hidden={mode !== 'auto'} onClick={handleCutPathByGap} type="default">
                {tabPanel.apply}
              </Button>
            </Form>
          </div>
        </div>
        <Footer>
          <Button key="ok" onClick={onComplete} type="primary">
            {global.save_and_exit}
          </Button>
        </Footer>
      </Flex>
    </FullWindowPanelSider>
  );
}

const Sider = memo(UnmemorizedSider);

export default Sider;
