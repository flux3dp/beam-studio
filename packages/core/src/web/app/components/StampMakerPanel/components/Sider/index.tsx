import React, { memo } from 'react';

import { Button, ConfigProvider, Flex } from 'antd';

import ImageEditPanelIcons from '@core/app/icons/image-edit-panel/ImageEditPanelIcons';
import BackButton from '@core/app/widgets/FullWindowPanel/BackButton';
import Footer from '@core/app/widgets/FullWindowPanel/Footer';
import Header from '@core/app/widgets/FullWindowPanel/Header';
import FullWindowPanelSider from '@core/app/widgets/FullWindowPanel/Sider';
import useI18n from '@core/helpers/useI18n';

import Content from './Content';
import styles from './index.module.scss';

interface Props {
  handleComplete: () => void;
  onClose: () => void;
}

function Sider({ handleComplete, onClose }: Props): React.JSX.Element {
  const { buttons: langButtons, global: langGlobal, image_edit_panel: lang } = useI18n();

  return (
    <ConfigProvider theme={{ components: { InputNumber: { controlWidth: 80 } } }}>
      <FullWindowPanelSider className={styles.sider}>
        <Flex className={styles['h-100']} justify="space-between" vertical>
          <div>
            <BackButton onClose={onClose}>{langButtons.back_to_beam_studio}</BackButton>
            <Header icon={<ImageEditPanelIcons.EditImage />} title={lang.title} />
            <Content />
          </div>
          <Footer>
            <Button key="ok" onClick={handleComplete} type="primary">
              {langGlobal.ok}
            </Button>
          </Footer>
        </Flex>
      </FullWindowPanelSider>
    </ConfigProvider>
  );
}

const MemorizedSider = memo(Sider);

export default MemorizedSider;
