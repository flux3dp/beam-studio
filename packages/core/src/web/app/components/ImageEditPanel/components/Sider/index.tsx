/* eslint-disable @typescript-eslint/no-shadow */
import React, { memo } from 'react';

import { Button, Flex, TabPaneProps, Tabs } from 'antd';
import useI18n from 'helpers/useI18n';

import ImageEditPanelIcons from 'app/icons/image-edit-panel/ImageEditPanelIcons';
import BackButton from 'app/widgets/FullWindowPanel/BackButton';
import Footer from 'app/widgets/FullWindowPanel/Footer';
import Header from 'app/widgets/FullWindowPanel/Header';
import FullWindowPanelSider from 'app/widgets/FullWindowPanel/Sider';

import Eraser from './Eraser';
import MagicWand from './MagicWand';

import styles from './index.module.scss';

interface Props {
  onClose: () => void;
  handleComplete: () => void;
  brushSize: number;
  setBrushSize: (size: number) => void;
  tolerance: number;
  setTolerance: (tolerance: number) => void;
  mode: 'eraser' | 'magicWand';
  setMode: (mode: 'eraser' | 'magicWand') => void;
  setOperation: (operation: 'eraser' | 'magicWand' | 'drag' | null) => void;
}

interface Tab extends Omit<TabPaneProps, 'tab'> {
  key: string;
  label: React.ReactNode;
}

function Sider({
  onClose,
  handleComplete,
  brushSize,
  setBrushSize,
  tolerance,
  setTolerance,
  mode,
  setMode,
  setOperation,
}: Props): JSX.Element {
  const {
    beambox: { photo_edit_panel: langPhoto },
    image_edit_panel: lang,
    buttons: langButtons,
  } = useI18n();

  const tabItems: Array<Tab> = [
    {
      key: 'eraser',
      label: lang.eraser.title,
      children: <Eraser brushSize={brushSize} setBrushSize={setBrushSize} />,
      icon: <ImageEditPanelIcons.Eraser />,
    },
    {
      key: 'magicWand',
      label: lang.magic_wand.title,
      children: <MagicWand tolerance={tolerance} setTolerance={setTolerance} />,
      icon: <ImageEditPanelIcons.MagicWand />,
    },
  ];

  return (
    <FullWindowPanelSider className={styles.sider}>
      <Flex className={styles['h-100']} vertical justify="space-between">
        <div>
          <BackButton onClose={onClose}>{langButtons.back_to_beam_studio}</BackButton>
          <Header icon={<ImageEditPanelIcons.EditImage />} title={lang.title} />
          <Tabs
            centered
            size="large"
            activeKey={mode}
            items={tabItems}
            onChange={(mode: 'eraser' | 'magicWand') => {
              setOperation(null);
              setMode(mode);
            }}
          />
        </div>
        <Footer>
          <Button key="ok" type="primary" onClick={handleComplete}>
            {langPhoto.okay}
          </Button>
        </Footer>
      </Flex>
    </FullWindowPanelSider>
  );
}

const MemorizedSider = memo(Sider);

export default MemorizedSider;
