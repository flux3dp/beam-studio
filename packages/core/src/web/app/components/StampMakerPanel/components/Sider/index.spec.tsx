import React from 'react';
import { fireEvent, render } from '@testing-library/react';

jest.mock('antd', () => ({
  ...jest.requireActual('antd'),
  ConfigProvider: ({ children, theme }: any) => (
    <div className="ant-config-provider" data-theme={JSON.stringify(theme)}>
      {children}
    </div>
  ),
}));


jest.mock('@core/app/icons/action-panel/ActionPanelIcons', () => ({
  __esModule: true,
  default: {
    Stamp: () => <svg data-testid="stamp-icon" />,
  },
}));

jest.mock('@core/app/widgets/FullWindowPanel/BackButton', () => ({ children, onClose }: any) => (
  <button data-testid="back-button" onClick={onClose}>
    {children}
  </button>
));

jest.mock('@core/app/widgets/FullWindowPanel/Footer', () => ({ children }: any) => (
  <div data-testid="footer">{children}</div>
));

jest.mock('@core/app/widgets/FullWindowPanel/Header', () => ({ icon, title }: any) => (
  <div data-testid="header">
    {icon}
    <span>{title}</span>
  </div>
));

jest.mock('@core/app/widgets/FullWindowPanel/Sider', () => ({ children, className }: any) => (
  <div className={className} data-testid="sider">
    {children}
  </div>
));

jest.mock('./Content', () => () => <div data-testid="sider-content">Content</div>);

import Sider from './index';

describe('test Sider', () => {
  const mockHandleComplete = jest.fn();
  const mockOnClose = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { container } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    expect(container).toMatchSnapshot();
  });

  it('should render header with icon and title', () => {
    const { getByTestId, getByText } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    expect(getByTestId('header')).toBeInTheDocument();
    expect(getByTestId('stamp-icon')).toBeInTheDocument();
    expect(getByText('Stamp Maker')).toBeInTheDocument();
  });

  it('should render back button with correct text', () => {
    const { getByTestId } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    const backButton = getByTestId('back-button');

    expect(backButton).toHaveTextContent('Back to Beam Studio');
  });

  it('should handle back button click', () => {
    const { getByTestId } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    const backButton = getByTestId('back-button');

    fireEvent.click(backButton);

    expect(mockOnClose).toHaveBeenCalled();
  });

  it('should render sider content', () => {
    const { getByTestId } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    expect(getByTestId('sider-content')).toBeInTheDocument();
  });

  it('should render OK button in footer', () => {
    const { getByText } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    const okButton = getByText('OK');

    expect(okButton).toBeInTheDocument();
    expect(okButton.closest('button')).toHaveClass('ant-btn-primary');
  });

  it('should handle OK button click', () => {
    const { getByText } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    const okButton = getByText('OK');

    fireEvent.click(okButton);

    expect(mockHandleComplete).toHaveBeenCalled();
  });

  it('should apply ConfigProvider theme', () => {
    const { container } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    const configProvider = container.querySelector('.ant-config-provider');

    expect(configProvider).toBeInTheDocument();
  });

  it('should apply correct className to sider', () => {
    const { getByTestId } = render(<Sider handleComplete={mockHandleComplete} onClose={mockOnClose} />);

    const sider = getByTestId('sider');

    expect(sider).toHaveClass('sider');
  });
});
