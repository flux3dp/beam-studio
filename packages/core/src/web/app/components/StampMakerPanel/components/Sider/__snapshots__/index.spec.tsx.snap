// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Sider should render correctly 1`] = `
<div>
  <div
    class="ant-config-provider"
    data-theme="{"components":{"InputNumber":{"controlWidth":80}}}"
  >
    <div
      class="sider"
      data-testid="sider"
    >
      <div
        class="h-100 ant-flex css-dev-only-do-not-override-hash ant-flex-align-stretch ant-flex-justify-space-between ant-flex-vertical"
      >
        <div>
          <button
            data-testid="back-button"
          >
            Back to Beam Studio
          </button>
          <div
            data-testid="header"
          >
            <svg
              data-testid="edit-image-icon"
            />
            <span>
              Stamp Maker
            </span>
          </div>
          <div
            data-testid="sider-content"
          >
            Content
          </div>
        </div>
        <div
          data-testid="footer"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-hash ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
            type="button"
          >
            <span>
              OK
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
